worker/src/RTC/RetransmissionBuffer.cpp

```diff
@@ -0,0 +1,530 @@
+				}
+
+				// We are done.
+				if (timestamp <= newerItem->timestamp)
+				{
+					break;
+				}
+				else
+				{
+					MS_WARN_TAG(
+					  rtp,
+					  "packet timestamp is higher than timestamp of immediate newer packet in the buffer, discarding it [ssrc:%" PRIu32
+					  ", seq:%" PRIu16 ", timestamp:%" PRIu32 "]",
+					  ssrc,
+					  seq,
+					  timestamp);
+
+					return;
+				}
+			}
+
+			// Store the packet.
+			item = new Item();
+
+			this->buffer[idx] = FillItem(item, packet, sharedPacket);
+		}
+
+#if MS_LOG_DEV_LEVEL == 3
+		Dump();
+#endif
+
+		MS_ASSERT(
+		  this->buffer.size() <= this->maxItems,
+		  "buffer contains %zu items (more than %" PRIu16 " max items)",
+		  this->buffer.size(),
+		  this->maxItems);
+	}
+
+	void RetransmissionBuffer::Clear()
+	{
+		MS_TRACE();
+
+		for (auto* item : this->buffer)
+		{
+			if (!item)
+			{
+				continue;
+			}
+
+			// Reset the stored item (decrease RTP packet shared pointer counter).
+			item->Reset();
+
+			delete item;
+		}
+
+		this->buffer.clear();
+		this->startSeq = 0u;
+	}
+
+	void RetransmissionBuffer::Dump() const
+	{
+		MS_TRACE();
+
+		MS_DUMP("<RetransmissionBuffer>");
+		MS_DUMP("  buffer [size:%zu, maxSize:%" PRIu16 "]", this->buffer.size(), this->maxItems);
+		if (this->buffer.size() > 0)
+		{
+			const auto* oldestItem = GetOldest();
+			const auto* newestItem = GetNewest();
+
+			MS_DUMP(
+			  "  oldest item [seq:%" PRIu16 ", timestamp:%" PRIu32 "]",
+			  oldestItem->sequenceNumber,
+			  oldestItem->timestamp);
+			MS_DUMP(
+			  "  newest item [seq:%" PRIu16 ", timestamp:%" PRIu32 "]",
+			  newestItem->sequenceNumber,
+			  newestItem->timestamp);
+			MS_DUMP(
+			  "  buffer window: %" PRIu32 "ms",
+			  static_cast<uint32_t>(newestItem->timestamp * 1000 / this->clockRate) -
+			    static_cast<uint32_t>(oldestItem->timestamp * 1000 / this->clockRate));
+		}
+		MS_DUMP("</RetransmissionBuffer>");
+	}
+
+	RetransmissionBuffer::Item* RetransmissionBuffer::GetOldest() const
+	{
+		MS_TRACE();
+
+		return this->Get(this->startSeq);
+	}
+
+	RetransmissionBuffer::Item* RetransmissionBuffer::GetNewest() const
+	{
+		MS_TRACE();
+
+		return this->Get(this->startSeq + this->buffer.size() - 1);
+	}
+
+	void RetransmissionBuffer::RemoveOldest()
+	{
+		MS_TRACE();
+
+		if (this->buffer.empty())
+		{
+			return;
+		}
+
+		auto* item = this->buffer.at(0);
+
+		// Reset the stored item (decrease RTP packet shared pointer counter).
+		item->Reset();
+
+		delete item;
+
+		this->buffer.pop_front();
+		this->startSeq++;
+
+		// Remove all nullptr elements from the beginning of the buffer.
+		// NOTE: Calling front on an empty container is undefined.
+		while (!this->buffer.empty() && this->buffer.front() == nullptr)
+		{
+			this->buffer.pop_front();
+			this->startSeq++;
+		}
+
+		// If we emptied the full buffer, reset startSeq.
+		if (this->buffer.empty())
+		{
+			this->startSeq = 0u;
+		}
+	}
+
+	void RetransmissionBuffer::RemoveFromFrontAtLeast(uint16_t numItems)
```

Comment:
for the same reason `RemoveOldest` is not called `RemoveOldestAtLeast` I would rename this method to `RemoveFromFront`

Reply:
What about RemoveOldest(numItems) instead? This is, a different signature

Reply:
I've done it and pushed it.